### PR TITLE
Add `Arithmetic` lint

### PR DESCRIPTION
**Pull request by c410-f3r**
Fixes https://github.com/rust-lang/rust-clippy/issues/8903

r? @llogiq 

changelog: Add `Arithmetic` lint

### CODE REVIEW

File: tests/ui/arithmetic.fixed [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    let saturating: Saturating<u32> = Saturating(0u32);
+    let wrapping: Wrapping<u32> = Wrapping(0u32);
+    let string: String = String::new();
+
+    let _ = saturating + saturating;
+    let _ = wrapping + wrapping;
+    let _ = string + "";
+}
+
+fn main() {}
```
Comment (c410-f3r) [inline tests/ui/arithmetic.fixed:23]: Not sure why this file is necessary

File: tests/ui/integer_arithmetic.stderr [hunk 2]
```diff
@@ -12,158 +12,182 @@ error: this operation will panic at runtime
...
+   |     ^^^^^^
+
+error: arithmetic detected
+  --> $DIR/integer_arithmetic.rs:30:5
+   |
+LL |     i /= 0;
+   |     ^^^^^^
+
+error: arithmetic detected
+  --> $DIR/integer_arithmetic.rs:31:5
```
Comment (c410-f3r) [inline tests/ui/integer_arithmetic.stderr:91]: Replacing `#![warn(clippy::integer_arithmetic, clippy::float_arithmetic)]` with `#![warn(clippy::arithmetic)]` actually improved diagnostics.

If you guys want to deprecate `integer_arithmetic` and `float_arithmetic` in favour of `arithmetic`, then I can scan for primitives to recover the lost "integer/float arithmetic detected" messages. Although that would imply more processing and such thing doesn't seem too useful at all.

File: rust-toolchain [hunk 3]
```diff
@@ -1,3 +1,3 @@
 [toolchain]
-channel = "nightly-2022-06-30"
-components = ["cargo", "llvm-tools-preview", "rust-src", "rust-std", "rustc", "rustc-dev", "rustfmt"]
+channel = "nightly-2022-06-29"
+components = ["llvm-tools-preview", "rust-src", "rust-std", "rustc", "rustc-dev", "rustfmt"]
```
Comment (c410-f3r) [inline rust-toolchain]: Please ignore this change. Local stuff...

File: clippy_lints/src/lib.rs [hunk 4]
```diff
@@ -547,6 +548,8 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
         store.register_late_pass(|| Box::new(utils::internal_lints::MsrvAttrImpl));
     }
 
+    let arithmetic_allowed = conf.arithmetic_allowed.clone();
+    store.register_late_pass(move || Box::new(arithmetic::Arithmetic::new(arithmetic_allowed.clone())));
```
Comment (c410-f3r) [inline clippy_lints/src/lib.rs:552]: Not sure how to avoid double clones

File: tests/ui/arithmetic.fixed (hunk 1 above)
Comment (c410-f3r) [inline tests/ui/arithmetic.fixed:23]: Not sure why this file is necessary

File: clippy_lints/src/arithmetic.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                } else if let hir::QPath::TypeRelative(ref ty, _) = qpath
+                    && let hir::TyKind::Path(hir::QPath::Resolved(_, ref path)) = ty.kind
+                {
+                    path
+                }
+                else {
+                    return false;
+                };
+                for segment in path.segments {
+                    if self.allowed.contains(segment.ident.as_str()) {
```
Comment (c410-f3r) [inline clippy_lints/src/arithmetic.rs:64]: `Ident::as_str` actually does some processing. Maybe use `FxHashSet<Ident>` instead?

File: clippy_lints/src/arithmetic.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+                        path
+                    }
+                } else if let hir::QPath::TypeRelative(ref ty, _) = qpath
+                    && let hir::TyKind::Path(hir::QPath::Resolved(_, ref path)) = ty.kind
+                {
+                    path
+                }
+                else {
+                    return false;
+                };
```
Comment (c410-f3r) [inline clippy_lints/src/arithmetic.rs:62]: Only catches

```rust
// Declared types
let local: Wrapping<u32> = Wrapping::new(0);
local + local

// Constructors
Wrapping::new(0) + Wrapping::new(0)
```

I.e., locals without explicit types aren't evaluated. Does `clippy_utils` has an utility to iterate a path in order to retrieved its inferred type?

```rust
let n = Wrapping::new(0);
let another = n;
another + another
```

File: tests/ui/integer_arithmetic.stderr (hunk 2 above)
Comment (Jarcho) [reply tests/ui/integer_arithmetic.stderr:91]: `integer_arithmetic` specifically checks for uses which may panic. `float_arithmetic` could maybe be done, but that can be a later problem.

File: clippy_lints/src/arithmetic.rs (hunk 5 above)
Comment (llogiq) [reply clippy_lints/src/arithmetic.rs]: At least pull the `as_str()` out of the loop.

**Comment by Jarcho** [comment]
Just a general comment about the approach here. This would be better done using resolved items rather than paths. `LateContext::qpath_res` will resolve any `QPath` if you need it. Generally you should be able to use `cx.typeck_results().expr_ty(_)` to get the type of an `Expr`. You can also look at the `disallowed_types` lint for how it deals with configuring which types to block.

**Comment by c410-f3r** [comment]
> Just a general comment about the approach here. This would be better done using resolved items rather than paths. `LateContext::qpath_res` will resolve any `QPath` if you need it. Generally you should be able to use `cx.typeck_results().expr_ty(_)` to get the type of an `Expr`. You can also look at the `disallowed_types` lint for how it deals with configuring which types to block.

Exactly what I was looking for! Thanks!

**Comment by c410-f3r** [comment]
I think the code is now in a good shape for review.

File: tests/ui/integer_arithmetic.stderr (hunk 2 above)
Comment (c410-f3r) [reply tests/ui/integer_arithmetic.stderr]: `i /= 0` and `i %= 0` will panic when the compiler is not able to deduct the RHS at compile-time so why the lint currently omits them?

Regardless, I reverted these changes for now but they can be reviewed in the future.

Comment (Jarcho) [reply tests/ui/integer_arithmetic.stderr]: Both `/= 0` and `%= 0` are caught by the `unconditional_panic` lint.

**Comment by c410-f3r** [comment]
@llogiq Perhaps another reviewer?

**Comment by llogiq** [comment]
No, I just had private stuff keeping me from reviewing. Sorry for the delay. I agree, the code is in good shape.

@bors r+

**Comment by c410-f3r** [comment]
> No, I just had private stuff keeping me from reviewing. Sorry for the delay. I agree, the code is in good shape.
> 
> @bors r+

Thank you @llogiq

**Pull request merged**